src/main.rs

```diff
@@ -0,0 +1,272 @@
+
+use argparse::{ArgumentParser, Store, StoreOption};
+use syslog::Facility;
+use wait_timeout::ChildExt;
+
+lazy_static! {
+    static ref TOTAL_MOUNTS: prometheus::Gauge = register_gauge!(
+        "total_mountpoints",
+        "Total number of mountpoints"
+    ).unwrap();
+
+    static ref DEAD_MOUNTS: prometheus::Gauge = register_gauge!(
+        "dead_mountpoints",
+        "Number of unresponsive mountpoints"
+    ).unwrap();
+}
+
+mod get_mounts;
+
+#[derive(Debug)]
+struct MountStatus {
+    last_checked: Instant,
+    alive: bool,
+    check_process: Option<process::Child>,
+}
+
+fn handle_syslog_error(err: std::io::Error) -> usize {
+    // Convenience function allowing all of our syslog calls to use .unwrap_or_else
+    eprintln!("Syslog failed: {}", err);
+    0
+}
+
+fn main() {
+    let mut poll_interval = 60;
+    let mut prometheus_push_gateway: Option<String> = None;
+
+    {
+        // this block limits scope of borrows by ap.refer() method
+        let mut ap = ArgumentParser::new();
+        ap.set_description(concat!(
+            "Monitor the status of mounted filesystems and report inaccessible mounts.",
+            " Dead mounts will be reported to the local syslog server and optionally",
+            " a Prometheus push-gateway service."
+        ));
+
+        ap.refer(&mut prometheus_push_gateway).add_option(
+            &["--prometheus-push-gateway"],
+            StoreOption,
+            "Location of the Prometheus push-gateway server to send metrics to",
+        );
+        ap.refer(&mut poll_interval).add_option(
+            &["--poll-interval"],
+            Store,
+            "Number of seconds to wait before checking mounts",
+        );
+        ap.parse_args_or_exit();
+    }
+
+    let poll_interval_duration = Duration::from_secs(poll_interval);
+
+    println!(
+        "mount_status_monitor checking mounts every {} seconds",
+        poll_interval_duration.as_secs()
+    );
+
+    let syslog = syslog::unix(Facility::LOG_DAEMON).unwrap_or_else(|err| {
+        eprintln!("Unable to connect to syslog: {}", err);
+        std::process::exit(1);
+    });
+
+    let mut mount_statuses = HashMap::<String, MountStatus>::new();
+
+    loop {
+        check_mounts(&mut mount_statuses, &syslog);
+
+        // We calculate these values each time because a filesystem may have been
+        // mounted or unmounted since the last check:
+        let total_mounts = mount_statuses.len();
+        let dead_mounts = mount_statuses
+            .iter()
+            .map(|(_, status)| !status.alive)
+            .filter(|&i| i)
+            .count();
+
+        // TODO: consider making this debug or sending it to stdout?
+        syslog
+            .info(format!(
+                "Checked {} mounts; {} are dead",
+                total_mounts,
+                dead_mounts
+            ))
+            .unwrap_or_else(handle_syslog_error);
+
+        if prometheus_push_gateway.is_some() {
```

Comment:
You can use `if let &Some(ref prometheus_push_gateway) = promethrues_push_gateway {` and avoid the calls to `as_ref` and `unwrap` below.

Reply:
Thanks – I was wondering about which was considered more idiomatic last night.